### ISSUE TITLE
SystemChrome.setPreferredOrientations does not force the device to the given orientations until the device is physically rotated

### ISSUE THREAD

**Issue by sroddy** [flutter/flutter#13238]
## Steps to Reproduce

Running on iOS 10/11

Start an app in portrait mode and invoke:
```dart
SystemChrome.setPreferredOrientations([ DeviceOrientation.portraitUp ]);
```

then invoke
```dart
SystemChrome.setPreferredOrientations([ DeviceOrientation.landscapeLeft ]);
```

Notice that neither the device neither the simulator automatically rotate until you physically rotate the device (and does not rotate at all if the device rotation trigger is locked).

This happens because calling +[UIViewController attemptRotationToDeviceOrientation] here https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm#L703 is unfortunately not enough.

I workarounded the issue creating a small platform channel that invokes this code for switching to portrait right before the call to setPreferredOrientations:
```objc
[[UIDevice currentDevice] setValue:@(UIInterfaceOrientationPortrait) forKey:@"orientation"];
```
And the counterpart code for switching to landscape
```objc
[[UIDevice currentDevice] setValue:@(UIInterfaceOrientationLandscapeLeft) forKey:@"orientation"];
```

This solved my issue.
I think the current behaviour under iOS is bugged, moreover because my expected behaviour is what is currently happening on the Android counterpart without the need of any workaroud.

I know that the proposed way to fix it looks bad (using key-value coding on a read-only property), but if you search around the Internet, it seems to be the most reliable (and possibly only) way to handle this thing when you have just one view / view controller.

**Comment by sroddy** [comment]
cc @chinmaygarde

**Comment by chinmaygarde** [comment]
I am not sure why this is not working for you. From my reading of the documentation on [`-[UIViewController supportedInterfaceOrientations]` and `-[UIViewController shouldAutorotate]`](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621435-supportedinterfaceorientations?language=objc), the responses that the `FlutterViewController` makes to UIKit seem to be correct (assuming that the view controller is the top most view controller in the view controller hierarchy).

Have you checked whether the supported interface orientations are present in the application's `Info.plist` and that the application or its delegate is not giving conflicting replies in the [`-[UIApplicationDelegate application:supportedInterfaceOrientationsForWindow:]`](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623107-application?language=objc) method?

**Comment by chinmaygarde** [comment]
cc @cbracken

**Comment by sroddy** [comment]
I know they seem right. Unfortunately they don't work.

https://stackoverflow.com/a/19125466

```
A call to attemptRotationToDeviceOrientation only rotates the orientation of the interface, if and only if, the device itself has been rotated.
```

I've been fighting in the past with this issue in the last 1 year and half and so far the only consistent way I've found to deal with this is to check the real device orientation and force it to one of the supported ones if unsupported.

**Comment by sroddy** [comment]
@chinmaygarde  @cbracken you can see here a basic example demonstrating the issue:

- create a new flutter project
- replace the contents on main.dart with this code: https://gist.github.com/sroddy/22ffeb04a542748fc141b322921f5f28
- start the project in a iOS device / simulator
- tap the + button to change the orientation of the interface

Notice that:
- If the previous device orientation was one of the supported ones (e.g. you keep the device in portraitUp mode or you lock the screen rotation), pressing the button won't produce any visible effect.
- if you set the orientation to landscapeLeft, the orientation will change as soon as you physically rotate the device (cmd-right-arrow in iOS simulator). if you now keep the device in landscapeLeft, tapping the button won't produce any visible effect.
- If you rotate the device in one of the two non-supported orientations (landscapeRight or portraitUp), tapping the button will produce the expected behaviour, causing the interface to rotate accordingly to the setting any time you tap on the button.
- If you run the same code in Android, the behaviour is as (I guess) expected.

**Comment by btastic** [comment]
We just noticed that when the device was in landscape and then the flutter app is opened with 

```dart
SystemChrome.setPreferredOrientations([
  DeviceOrientation.portraitUp,
]);
```

The app will open in landscaped regardless of this setting.

**Comment by CaiJingLong** [comment]
In iPad, this method doesn't seem to be implemented.
When I rotated iPad, flutter's UI rotated.

**Comment by miguelpruivo** [comment]
+1

**Comment by crnaosa** [comment]
+1

**Comment by zoechi** [comment]
@miguelpruivo @crnaosa 
Using "add reaction" on the initial comment would increase priority while `+1` comments are rather pointless and cause lots of people to get notified for no reason.
If you want to get notified just click the "Subscribe" button to the right.

**Comment by crnaosa** [comment]
@sroddy thanks for opening this thread. It's almost as I felt autistic for having this problem. Since I am android developer trying to create a flutter app, I am not really sure where should i put the objective c code.
Any help would be much appreciated!
@zoechi Ok, I'll do that next time. Any help to workaround this would be much appreciated.

**Comment by lennartschoch** [comment]
I am experiencing the same issue both with the iOS emulator (iPad Pro 9.7, iOS 12) and a physical device (iPad Air 2, iOS 11). I implemented the setPreferredOrientations method inside the main function, but there is no effect noticeable. This is a huge problem when developing with user interfaces which are designed to only work in either portrait or landscape mode, not both. Can we get an update on this?

**Comment by albo1337** [comment]
@sroddy Could you please share your solution? I'm facing the same issue and have no idea how to fix this on iOS. It seems that you fixed it.

**Comment by lennartschoch** [comment]
@albo1337 Not an ideal solution, but if you just want to get your whole app locked to a certain orientation, you can specify that inside the Info.plist file inside `ios/Runner`:

```
<key>UISupportedInterfaceOrientations</key>  
<array>
    <string>UIInterfaceOrientationPortrait</string
    <string>UIInterfaceOrientationLandscapeLeft</string
    <string>UIInterfaceOrientationLandscapeRight</string>
</array>
<key>UISupportedInterfaceOrientations~ipad</key>
<array>
    <string>UIInterfaceOrientationPortrait</string
    <string>UIInterfaceOrientationPortraitUpsideDown</string
    <string>UIInterfaceOrientationLandscapeLeft</string
    <string>UIInterfaceOrientationLandscapeRight</string>
</array>
```

Just remove the orientations that you do not support.

**Comment by albo1337** [comment]
@lennartschoch the problem is, I need the landscape mode only for one screen and the portrait mode for every other screen... Your solution does not help me with my current problem, but thank you a lot for your help!

**Comment by miguelpruivo** [comment]
@albo1337 that’s exactly what I need as well.

**Comment by jaumard** [comment]
@albo1337 I'm in the same case, I need the device in landscape for one screen and all the other in portrait. I have the same issue as you, you need to rotate the device for the UI to follow :( 
Works great on Android, but not on iOS

**Comment by sroddy** [comment]
@albo1337 You basically need to create a custom platform channel and use it to force the desired orientation right after setting the preferred one. In Dart:
```dart
SystemChrome.setPreferredOrientations([ DeviceOrientation.landscapeLeft ]);
_rotationChannel.invokeMethod('setLandscape');
```

in iOS code:
```
[[UIDevice currentDevice] setValue:@(UIInterfaceOrientationLandscapeLeft) forKey:@"orientation"];
```

**Comment by GaryQian** [comment]
cc @chinmaygarde Does this new information bring anything to light on this issue?

**Comment by eseidelGoogle** [comment]
This sounds like it might be an easy fix?  Maybe a starter bug for one of the new engine hackers @cbracken?

The implementation of SystemChrome is here:
https://github.com/flutter/engine/search?q=systemChrome&unscoped_q=systemChrome

**Comment by jaumard** [comment]
Ha ha @eseidelGoogle ! Easy fix when you know where to do it 😆 I've searched on the source where to do this fix with no luck ^^ thanks for pointing that out ! 
But for now I'm using @sroddy 's solution and force screen rotation by myself until the fix :)

**Comment by eseidelGoogle** [comment]
Looks like someone even wrote a medium article about this and how to work around it. :/
https://medium.com/@kr1uz/how-to-restrict-device-orientation-in-flutter-65431cd35113

**Comment by eseidelGoogle** [comment]
CC @kriuz the author of said article.

**Comment by lennartschoch** [comment]
@eseidel I actually tried exactly that and couldn't get the orientation locked on an iPad Air 2 :/

**Comment by eseidelGoogle** [comment]
It's hard for me to parse from this issue how much of this could be Flutter itself, vs if the plugins/APIs involved are just not calling the right iOS APIs to implement their desired functionality.  Either are probably easy to fix?

@xster might have a guess.

**Comment by eseidelGoogle** [comment]
Patches to the engine would certainly be welcome:
https://github.com/flutter/engine/search?q=systemChrome&unscoped_q=systemChrome :)

**Comment by xster** [comment]
I think the issue is the discrepancy between the names and docs of [SystemChrome.setPreferredOrientations](https://docs.flutter.io/flutter/services/SystemChrome/setPreferredOrientations.html) in Flutter, [Activity.setRequestedOrientation](https://developer.android.com/reference/android/app/Activity#setRequestedOrientation(int)) in Android and [UIViewController.supportedInterfaceOrientations](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621435-supportedinterfaceorientations?language=objc) in iOS. 

Arguably, Android is the odd one out since `setPreferredOrientations` is documented as:

```
Specifies the set of orientations the application interface can be displayed in.
```

and `setRequestedOrientation` actually performs the non-conforming

```
If the activity is currently in the foreground or otherwise impacting the screen orientation, the screen will immediately be changed
```

If anyone is willing to send us a PR, I think we should do the following:

On `setPreferredOrientations`, 

- on iOS, we're arguably already doing the right thing based on the Dart side doc. Though we should force a rotation if the current device orientation is no longer in the list of allowed `orientations`.
- on Android, our present behavior isn't right. We should also |= in ActivityInfo.SCREEN_ORIENTATION_SENSOR if the user's preferences has that set. We should also likely remove [this logic](https://github.com/flutter/engine/blob/4b0f2cac11c7932825d08866761a5c150db194fe/shell/platform/android/io/flutter/plugin/platform/PlatformPlugin.java#L161) and just let the orientation default to ActivityInfo.SCREEN_ORIENTATION_USER if the present orientation is part of the `orientations` list.

We should also create a new `SystemChrome.forceOrientation(DeviceOrientation orientation)` which

- on iOS, uses the mechanism @sroddy described.
- on Android, calls `setRequestedOrientation` with the added flags ActivityInfo.SCREEN_ORIENTATION_LOCKED and ActivityInfo.SCREEN_ORIENTATION_NOSENSOR.

**Comment by arturomf** [comment]
It would be possible to get more information in how to implement @sroddy solution? i have tried reading info about  platform channels but i'm a bit new with flutter...

Thanks in advance!

**Comment by sfaujour** [comment]
@arturomf Hey, i've published a package https://pub.dartlang.org/packages/auto_orientation to set the orientations and force the device programmatically to rotate.

**Comment by arturomf** [comment]
Hi @sfaujour! i have tested your package but it's not working in an iPad. This is the code i'm using:

```
@override
  void initState() {
    super.initState();

    SystemChrome.setEnabledSystemUIOverlays([]);

    SystemChrome.setPreferredOrientations([
      DeviceOrientation.landscapeRight,
      DeviceOrientation.landscapeLeft,
    ]);

    AutoOrientation.landscapeRightMode();
}
```

I'm using only in one of my app screens, but when i navigate to it in portrait mode, it does not rotate to landscape. Any idea?

Thanks!

**Comment by sfaujour** [comment]
@arturomf I've to say.. i really don't know. I will get deeper into the issue.. thats really strange.

**Comment by arturomf** [comment]
@sfaujour Thanks in advance, if i can help with testing or more info i will be delighted to do it

**Comment by sfaujour** [comment]
@arturomf I think this miss behavior is related to this issue, described in https://github.com/flutter/flutter/issues/27235

**Comment by arturomf** [comment]
Hi all, i have implemented @sroddy solution with a platform channel, but it seems to work only in iphone, not in iPad. Anyone have had success with this platform channel solution in iPad?

And another question.. when i use the platform channel method, the orientation is locked in iOS, how can i unlock it, so when i navigate out of the landscape screen, the orientation can be anyone.

Thanks!

**Comment by dnfield** [comment]
On iOS you can use https://pub.dartlang.org/packages/native_device_orientation

**Comment by vernvn** [comment]
iphone is ok , but ipad is not working,

**Comment by jaumard** [comment]
Any update on this ? would be nice to have this fixed

**Comment by besserwisser** [comment]
Would be really nice. Currently exiting a full screen video player in landscape shows an unwanted landscape app (I just want to show it in portrait)...

**Comment by liuyasss** [comment]
> @arturomf Hey, i've published a package https://pub.dartlang.org/packages/auto_orientation to set the orientations and force the device programmatically to rotate. I think this is the one you need..

not work on ios

**Comment by sfaujour** [comment]
> > @arturomf Hey, i've published a package https://pub.dartlang.org/packages/auto_orientation to set the orientations and force the device programmatically to rotate. I think this is the one you need..
> 
> not work on ios

Sure it works on iOS - it's currently not working for iPad's because of this problem https://github.com/flutter/flutter/issues/27235

**Comment by liuyasss** [comment]
> > > @arturomf Hey, i've published a package https://pub.dartlang.org/packages/auto_orientation to set the orientations and force the device programmatically to rotate. I think this is the one you need..
> > 
> > 
> > not work on ios
> 
> Sure it works on iOS - it's currently not working for iPad's because of this problem #27235

no , I get this package on my ios phone ,but not work ,

**Comment by tapizquent** [comment]
Is there any updates on this? Was marked as an easy-fix long time ago :(

It is really frustrating when need your app to rotate along with the UI on an action.

**Comment by elmariocarlos** [comment]
> @albo1337 Not an ideal solution, but if you just want to get your whole app locked to a certain orientation, you can specify that inside the Info.plist file inside `ios/Runner`:
> 
> ```plist
> <key>UISupportedInterfaceOrientations</key>  
> <array>
>     <string>UIInterfaceOrientationPortrait</string
>     <string>UIInterfaceOrientationLandscapeLeft</string
>     <string>UIInterfaceOrientationLandscapeRight</string>
> </array>
> <key>UISupportedInterfaceOrientations~ipad</key>
> <array>
>     <string>UIInterfaceOrientationPortrait</string
>     <string>UIInterfaceOrientationPortraitUpsideDown</string
>     <string>UIInterfaceOrientationLandscapeLeft</string
>     <string>UIInterfaceOrientationLandscapeRight</string>
> </array>
> ```
> 
> Just remove the orientations that you do not support.
> 
> Edit: Of course, this doesn't help if you need to lock specific views of your app, but might be helpful for anyone who stumbles upon this thread and wants their whole app locked ;)

Hi @lennartschoch , I have an app that will be locked to one orientation; for iOS this would work. Any idea how to get the lock in for Android? Thanks in advance.

**Comment by josh-ksr** [comment]
@eseidelGoogle I created a pull request to fix the iOS-problem and force an orientation change when the current interface orientation is not in the list of preferredOrientations [here](https://github.com/flutter/engine/pull/12295)

**Comment by escamoteur** [comment]
This might be another example for the same problem https://github.com/flutter/flutter/issues/40745

**Comment by alexei-g-aloteq** [comment]
This issue definitely need to be fixed ASAP. Makes Flutter look bad in sense of production platform...

**Comment by mofada** [comment]
When can solve this problem? It’s almost two years since I opened it in 2017. My God.

**Comment by GaryQian** [comment]
@josh-ksr's PR has landed to fix this for iOS, Still looking for an Android equivalent.

**Comment by sroddy** [comment]
@GaryQian it's been a while (!) since I reported this, but as far as I can remember the Android counterpart was already working as expected

**Comment by mofada** [comment]
@GaryQian Yes, the Android part works fine

**Comment by sroddy** [comment]
although we are probably still missing the enhancements proposed here
https://github.com/flutter/flutter/issues/13238#issuecomment-457001588

**Comment by elmariocarlos** [comment]
Will you confirm @GaryQian as soon as it is implemented? To ensure this thread is properly closed.

**Comment by lukasgit** [comment]
@eseidelGoogle @GaryQian we're part of the Google Cloud Startup program and also waiting on this iPad issue to be resolved before we can move to production. Thanks!

**Comment by lukasgit** [comment]
@timsneath ^^

**Comment by Hixie** [comment]
We believe this is fixed in `dev`. If you were running into this before, can you try `flutter channel dev; flutter upgrade` and see if you still see this?

**Comment by josh-ksr** [comment]
@Hixie i don´t think it is fixed in dev as the merge was reverted. There´s a new merge request (https://github.com/flutter/engine/pull/13170) but that´s not merged yet.

**Comment by mav2287** [comment]
I just tried the Dev channel and it is still not working as expected.

**Comment by GaryQian** [comment]
@josh-ksr Sorry for the delay, it is merged again!

**Comment by GaryQian** [comment]
Change is in master again as of https://github.com/flutter/flutter/commit/353721a1665c7b5d28c9ff4b9ff1c38e2f1a7d03

**Comment by elmariocarlos** [comment]
Awesome news! @sroddy @btastic @lukasgit would you test for implementation?

**Comment by cbracken** [comment]
I've removed the platform-android label, since this issue is very embedder-specific and the original issue, workarounds in comments, and fix are all iOS-specific. If people are running into a similar issue on Android, please file an issue and link it here from a comment to help others find it and add a thumbs-up reaction.

Just so @elmariocarlos 's question does't get lost after this comment: @sroddy @btastic @lukasgit I'm closing this issue as fixed now that the patch has landed, but please let us know if you're still seeing this and we'll re-open.

**Comment by kanthi0802** [comment]
Issues still exists in both Andorid and iOS. Below is the Flutter Doctor report 


[✓] Flutter (Channel master, v1.10.16-pre.67, on Mac OS X 10.14.1 18B75, locale en-GB)
    • Flutter version 1.10.16-pre.67 at /Users/kanthis/Documents/Flutter/flutter
    • Framework revision e3742b6e41 (5 hours ago), 2019-11-06 23:13:17 -0500
    • Engine revision c448d39d53
    • Dart version 2.7.0

[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/kanthis/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[!] Xcode - develop for iOS and macOS (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    ✗ Flutter requires a minimum Xcode version of 10.2.0.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.6.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 41.0.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (3 available)
    • Android SDK built for x86 • emulator-5554 • android-x86    • Android 9 (API 28) (emulator)
    • Chrome                    • chrome        • web-javascript • Google Chrome 78.0.3904.87
    • Web Server                • web-server    • web-javascript • Flutter Tools

! Doctor found issues in 2 categories.

**Comment by mofada** [comment]
I tried it, it is ok,

`Channel master, v1.10.16-pre.88`

**Comment by mofada** [comment]
@kanthi0802

**Comment by mofada** [comment]
upgrade your flutter and retry

**Comment by kanthi0802** [comment]
Not working for me tried with
Flutter (Channel master, v1.10.16-pre.110, on Mac OS X 10.14.1 18B75, locale en-GB).
Should we do anything other than calling  
**SystemChrome.setPreferredOrientations(
        [DeviceOrientation.landscapeLeft, DeviceOrientation.landscapeRight]);** ?

**Comment by ztpr0035** [comment]
> @lennartschoch the problem is, I need the landscape mode only for one screen and the portrait mode for every other screen... Your solution does not help me with my current problem, but thank you a lot for your help!

Hello @lennartschoch, I think this will solve your issue , I tried it and it work on both platforms,
return RotatedBox(
  quarterTurns: 1,
    child: Scaffold(),);

**Comment by cbracken** [comment]
Sounds like this is resolved based on other reports.

@lennartschoch please let us know if @ztpr0035's suggestion doesn't work for you.

**Comment by mirkopoloni** [comment]
Working fine on iPhone but still having problems on iPad, am I the only one?

`Channel master, v1.13.4-pre.12, on Mac OS X 10.15.1`

**Comment by darren-f5s** [comment]
Oh man, I'm forced to use v1.9.x due to older phone compatability... Any workaround for that version?

**Comment by freestyletime** [comment]
Hi, the library is unuseful to orientate the screen on androidv4.4.2

**Comment by waitwalker** [comment]
Some features of Flutter are not iOS friendly. For example, launch image, screen rotation, etc.

**Comment by hawkinsjb1** [comment]
Is the only solution to this really to use a RotatedBox as @ztpr0035 mentioned? This doesn't properly change the screen's size like a native orientation change. I'm having lots of overflow errors trying to use that solution

**Comment by pisces116** [comment]
Why this issue is not resolved but is marked as resolved

**Comment by SaidiAli** [comment]
The issue is closed but I think i can add this just incase. I am just learning flutter and in a lecture by Maximiliano of academind, he points out that for the issue at hand to work you need to add a `WidgetFlutterBinding.ensureInitialised()`.
Like:
`WidgetFlutterBinding.ensureInitialised();
SystemChrome.setPrefferedOrientation();`

**Comment by yubaokang** [comment]
> 在iPhone上可以正常工作，但在iPad上仍然有问题，我是唯一的一个吗？
> 
> `Channel master, v1.13.4-pre.12, on Mac OS X 10.15.1`
> 
> 1月13日更新
> 我刚刚发现，如果没有一些限制，这实际上无法在iPad上运行，这不是bug。...抱歉，如果我在写评论之前没有找到它。
> 作为其他人的参考，请在此处获取更多信息：
> [＃27235（评论）](https://github.com/flutter/flutter/issues/27235#issuecomment-530726236)
> https://api.flutter.dev/flutter/services/SystemChrome/setPreferredOrientations.html

then same
